include/irr/asset/IVirtualTexture.h

```diff
@@ -0,0 +1,750 @@
+#ifndef __IRR_I_VIRTUAL_TEXTURE_H_INCLUDED__
+#define __IRR_I_VIRTUAL_TEXTURE_H_INCLUDED__
+
+#include "irr/asset/format/EFormat.h"
+#include "irr/core/alloc/GeneralpurposeAddressAllocator.h"
+#include "irr/core/alloc/PoolAddressAllocator.h"
+#include "irr/core/math/morton.h"
+#include "irr/core/alloc/address_allocator_traits.h"
+#include "irr/core/memory/memory.h"
+#include "irr/asset/filters/CPaddedCopyImageFilter.h"
+#include "irr/asset/filters/CFillImageFilter.h"
+
+namespace irr {
+namespace asset
+{
+
+template <typename image_view_t>
+class IVirtualTexture : public core::IReferenceCounted
+{
+protected:
+    //! SPhysPgOffset is what is stored in texels of page table!
+    struct SPhysPgOffset
+    {
+        _IRR_STATIC_INLINE_CONSTEXPR uint32_t invalid_addr = 0xffffu;
+
+        SPhysPgOffset(uint32_t _addr) : addr(_addr) {}
+
+        //upper 16 bits are used for address of mip-tail page
+        uint32_t addr;
+    };
+
+    uint32_t neededPageCountForSide(uint32_t _sideExtent, uint32_t _level)
+    {
+        return (std::max(_sideExtent>>_level, 1u)+m_pgSzxy-1u) / m_pgSzxy;
```

Comment:
This is wrong, example... texture with a sidelength of 1025, gets padded to 2048 (with some wrap modes)
Mip Level 0: uses 1025 pixels out of 2048, with a page size of 128 it needs 9 pages
Mip Level 1: uses 513 pixels out of 1024 (coordinate 1.0 will map to 512.5 , the mip is computed from padded+wrapped image), it needs 5 pages
Mip Level 2: uses 257 pixels out of 512 (1.0-> 256.25), it needs 3 pages
Mip Level 3: uses 129 pixels out of 256, it needs 2 pages
Mip Level 4: uses 65 pixels out of 128, it needs 1 page
rest needs 1 page

Your function comes out as, 9 pages, then 4,2,1,......

You should swap`std::max(_sideExtent>>_level, 1u)` which rounds down

to `((_sideExtent+(0x1u<<_level)-1u)>>_level)` which rounds up and doesn't need a MAX

Reply:
miplevel `N` size of texture sized `S` in some dimension has size of `max(S / 2^N,1)` which is `max(S>>N,1)` so i think everything's fine
(so miplevel 1 of texture sized 1025 has size of 512, not 513)

Reply:
nope. because we pad the thing!